### PR TITLE
new Item tag/category relations and APIs

### PR DESCRIPTION
**Pull request by yiyione**
new relations:
ItemTagRelation / ItemCategoryRelation

image

New APIs

router
  .route('/items/:itemId/tag')
  .get(token.checkAuthAndGetUserInfo, itemController.listTags); // list item's all tags

router
  .route('/items/:itemId/tag/:tagId')
  .post(token.checkAuthAndGetUserInfo, itemController.addTag) // add tag to item
  .delete(token.checkAuthAndGetUserInfo, itemController.deleteTag); // delete tag from item

router
  .route('/items/:itemId/category')
  .get(token.checkAuthAndGetUserInfo, itemController.listCategories); // list item's all categories

router
  .route('/items/:itemId/category/:categoryId')
  .post(token.checkAuthAndGetUserInfo, itemController.addCategory) // add category to item
  .delete(token.checkAuthAndGetUserInfo, itemController.deleteCategory); // delete category from item

### CODE REVIEW

File: rest_server/src/models/model_init_handler.js [hunk 1]
```diff
@@ -46,12 +49,40 @@ const createTemplates = async models => {
...
         };
-        await models.MarketplaceItem.orm.create(newItem);
+        const marketplaceItem = await models.MarketplaceItem.orm.create(
+          newItem,
+        );
+        if (Array.isArray(newItem.categories)) {
+          for (const categoryName of newItem.categories) {
+            itemCategories.push([categoryName, marketplaceItem.id]);
+          }
+        }
```
Comment (hippogr) [inline rest_server/src/models/model_init_handler.js:59]: and else?

File: rest_server/src/models/model_init_handler.js [hunk 2]
```diff
@@ -46,12 +49,40 @@ const createTemplates = async models => {
...
+      category.addMarketplaceItem(itemId);
+    }
+    for (const [tagName, itemId] of itemTags) {
+      const [tag] = await models.ItemTag.orm.findOrCreate({
+        where: { name: tagName },
+      });
+      tag.addMarketplaceItem(itemId);
+    }
+  } catch (err) {
+    console.log(err.message);
```
Comment (hippogr) [inline rest_server/src/models/model_init_handler.js:84]: >console [](http://example.com/codeflow?start=4&length=7)

just curious: if some of the items are added successfully, and then an exception happens, what should we do? we need rollback or just keep the current state?

File: rest_server/src/models/marketplace_item.js [hunk 3]
```diff
@@ -165,11 +161,57 @@ class MarketplaceItem {
...
-        const items = await this.orm.findAll({ where: filterStatement });
+        const havings = [];
+        if (tags) {
+          havings.push({
+            where: where(
+              fn('array_agg', col('ItemTags.name')),
+              Op.contains,
+              cast(tags, 'varchar[]'),
+            ),
+          });
```
Comment (hippogr) [inline rest_server/src/models/marketplace_item.js:172]: looks like the processing logic for Item and ItemCatogory are similiar, so how about create a function to do that so you don't need repeat yourself?

File: rest_server/src/models/marketplace_item.js [hunk 4]
```diff
@@ -165,11 +161,57 @@ class MarketplaceItem {
...
+              model: this.models.ItemCategory.orm,
+              through: { attributes: [] },
+            },
+          ],
+        });
         return items;
       },
     );
-
     return await handler(
```
Comment (hippogr) [inline rest_server/src/models/marketplace_item.js:215]: er, why deleting this empty line? I quite like it to make the code not such busy...

File: rest_server/src/controllers/item_controller.js [hunk 5]
```diff
@@ -155,11 +155,152 @@ const del = asyncHandler(async (req, res, next) => {
...
+  }
+});
+
+const addTag = asyncHandler(async (req, res, next) => {
+  try {
+    let result = await MarketplaceItem.get(req.params.itemId);
+    if (checkReadPermission(req.userInfo, result)) {
+      result = await MarketplaceItem.addTag(result, req.params.tagId);
+      if (isNil(result)) {
+        return next(error.createNotFound());
```
Comment (hippogr) [inline rest_server/src/controllers/item_controller.js:186]: >createNotFound [](http://example.com/codeflow?start=26&length=14)

for a add function, NotFound is a little weird to me, but if you have your reasons, please ignore this comment. :-)

File: rest_server/src/models/model_init_handler.js (hunk 1 above)
Comment (yiyione) [reply rest_server/src/models/model_init_handler.js:62]: Added.

File: rest_server/src/models/model_init_handler.js (hunk 2 above)
Comment (yiyione) [reply rest_server/src/models/model_init_handler.js:90]: I think it's OK to keep the current state here.

File: rest_server/src/models/marketplace_item.js (hunk 3 above)
Comment (yiyione) [reply rest_server/src/models/marketplace_item.js]: Updated, add a new function and removed the duplicate code.

**Review by hippogr — APPROVED** [review]
:shipit:

**Review by debuggy — APPROVED** [review]

**Pull request merged**